cassandra/src/main/scala/geotrellis/spark/io/cassandra/CassandraCollectionReader.scala

```diff
@@ -58,18 +63,59 @@ object CassandraCollectionReader {
     else
       queryKeyBounds.flatMap(decomposeBounds)
 
-    val query = QueryBuilder.select("value")
-      .from(keyspace, table)
-      .where(eqs("key", QueryBuilder.bindMarker()))
-      .and(eqs("name", layerId.name))
-      .and(eqs("zoom", layerId.zoom))
-      .toString
+    val query = instance.cassandraConfig.partitionStrategy match {
+      case conf.`Read-Optimized-Partitioner` =>
+        QueryBuilder.select("value")
+          .from(keyspace, table)
+          .where(eqs("name", layerId.name))
+          .and(eqs("zoom", layerId.zoom))
+          .and(eqs("zoombin", QueryBuilder.bindMarker()))
+          .and(eqs("key", QueryBuilder.bindMarker()))
+          .toString
+
+      case conf.`Write-Optimized-Partitioner` =>
+        QueryBuilder.select("value")
+          .from(keyspace, table)
+          .where(eqs("key", QueryBuilder.bindMarker()))
+          .and(eqs("name", layerId.name))
+          .and(eqs("zoom", layerId.zoom))
+          .toString
+    }
+
+    lazy val intervals: Vector[(Interval[BigInt], Int)] = {
+      val ranges = keyIndex.indexRanges(keyIndex.keyBounds)
+
+      val binRanges = ranges.toVector.map{ range =>
+        val vb = new VectorBuilder[Interval[BigInt]]()
+        cfor(range._1)(_ <= range._2, _ + instance.cassandraConfig.tilesPerPartition){ i =>
+          vb += Interval.openUpper(i, i + instance.cassandraConfig.tilesPerPartition)
+        }
+        vb.result()
+      }
+
+      binRanges.flatten.zipWithIndex
+    }
+
+    def zoomBin(key: BigInteger): java.lang.Integer = {
```

Comment:
@ALPSMAC look into the tests here: https://github.com/locationtech/sfcurve i hope it's better than just words here under the issue. Look into the tests. I don't think we need `zoombin` at all.

SFC gives us a consistent enough ranges => i think we can use just this:

```scala
instance.cassandraConfig.partitionStrategy match {
        case conf.RadOptimizedPartitioner =>
          session.execute(
            SchemaBuilder.createTable(keyspace, table).ifNotExists()
              .addPartitionKey("name", text)
              .addPartitionKey("zoom", cint)
              .addClusteringColumn("key", varint)
              .addColumn("value", blob)
          )
        case conf.WriteOptimizedPartitioner =>
          session.execute(
            SchemaBuilder.createTable(keyspace, table).ifNotExists()
              .addPartitionKey("key", varint)
              .addClusteringColumn("name", text)
              .addClusteringColumn("zoom", cint)
              .addColumn("value", blob)
          )
      }
```

Reply:
Hey @pomadchin - thanks for taking the time to review what I've got here so far.  I appreciate your feedback!

It would be great news if, as you suggest, we could avoid the API change... this PR as it sits feels far more invasive than I'd like it to be.  I'll poke around more in the `sfcurve` repo as you suggested and see if the light-bulb flickers for me, but on first blush I'm not sure I follow your logic.

I don't want to jump to conclusions here - so please forgive my ignorance if I've grossly missed the mark, but... 

Wouldn't creating the `ReadOptimizedPartitioner` schema as you suggested here result in large partition sizes for higher zoom levels irrespective of the clustering column?  The partition size, as far as I know, is a function of the uniqueness of the partition keys - so the more collisions of a partition key you have the larger the partition grows.  In your example schema the partition size wouldn't have anything to do with the ranges of the SFC since that key is declared as only a clustering column (which determines order within a partition, but not the partition itself).

For instance - if I have a layer called "cloud_cover" and was indexing it at zoom level 15, wouldn't that mean that all of the tiles for zoom 15 for my cloud_cover layer would be stored in the same partition in Cassandra?  That seems like an unrealistically large partition, especially if I'm storing large tiles or double array tiles that don't compress well... 

Now - I am not at all a Cassandra expert (so perhaps I am completely out to lunch here), but Cassandra does warn about large partition sizes in its logs when they occur, and StackOverflow seems to agree (not that it's an authoritative source... but it's not a bad place to start): https://stackoverflow.com/questions/46272571/why-is-it-so-bad-to-have-large-partitions-in-cassandra

My premise in this PR was to give the client, if they so choose, an ability to tweak maximum partition sizes in a reasonable way to optimize performance for their given use case (with careful bench-marking).  How would indexing as you suggest avoid unrealistically large partition sizes (which would seem to me to break that premise at higher zoom levels)?

Reply:
Ah it makes sense, I was just confused about what it means; you answer clarifies everything. 

Sending you [here](https://github.com/locationtech/geotrellis/pull/2855/files#r248290447) than.

Reply:
Roger - thank you for the clarification :-) 

So I'll proceed with code cleanup then as well as working up some benchmarks to either prove or disprove the utility of this PR.